buffer/src/main/java/io/netty/buffer/api/internal/LeakDetection.java

```diff
@@ -0,0 +1,135 @@
+import io.netty.buffer.api.MemoryManager.LeakCallbackUninstall;
+import io.netty.util.internal.SystemPropertyUtil;
+import io.netty.util.internal.UnstableApi;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Utility class for the leak detection parts that are static and shared system-wide.
+ */
+@UnstableApi
+public final class LeakDetection {
+    static volatile int leakDetectionEnabled;
+
+    // Protected by synchronizing on the instance.
+    // This field is only accessed when leak are detected, or when callbacks are installed or removed.
+    private static final Set<Consumer<LeakInfo>> CALLBACKS = Collections.newSetFromMap(new IdentityHashMap<>());
+
+    private static final VarHandle LEAK_DETECTION_ENABLED_UPDATER;
+    static {
+        int enabled = SystemPropertyUtil.getBoolean("io.netty.buffer.leakDetectionEnabled", false) ? 1 : 0;
+        try {
+            LEAK_DETECTION_ENABLED_UPDATER = MethodHandles.lookup().findStaticVarHandle(
+                    LeakDetection.class, "leakDetectionEnabled", int.class);
+        } catch (Exception e) {
+            throw new ExceptionInInitializerError(e);
+        }
+        if (enabled > 0) {
+            CALLBACKS.add(LoggingLeakCallback.getInstance());
+        }
+        leakDetectionEnabled = enabled;
+    }
+
+    private LeakDetection() {
+    }
+
+    public static LeakCallbackUninstall onLeakDetected(Consumer<LeakInfo> callback) {
+        requireNonNull(callback, "callback");
+        LEAK_DETECTION_ENABLED_UPDATER.getAndAddAcquire(1);
+        synchronized (CALLBACKS) {
+            CALLBACKS.add(callback);
+        }
+        return new CallbackRemover(callback);
+    }
+
+    public static void reportLeak(LifecycleTracer tracer, String leakedObjectDescription) {
+        requireNonNull(tracer, "tracer");
+        requireNonNull(leakedObjectDescription, "leakedObjectDescription");
+        synchronized (CALLBACKS) {
+            if (!CALLBACKS.isEmpty()) {
+                LeakInfo info = new InternalLeakInfo(tracer, leakedObjectDescription);
+                for (Consumer<LeakInfo> callback : CALLBACKS) {
+                    callback.accept(info);
+                }
+            }
+        }
+    }
+
+    private static final class CallbackRemover implements LeakCallbackUninstall {
+        private final Consumer<LeakInfo> callback;
+
+        CallbackRemover(Consumer<LeakInfo> callback) {
+            this.callback = callback;
+        }
+
+        @Override
+        public void close() {
+            synchronized (CALLBACKS) {
+                CALLBACKS.remove(callback);
+            }
+            LEAK_DETECTION_ENABLED_UPDATER.getAndAddRelease(-1);
```

Comment:
we need to guard against the user calling `close()` multiple times and so have the updater changed multiple times. Also we want to guard the remove(...) call as well as the user might add the same instance again after it was removed and only the new returned `CallbackRemover` should be able to change the Set